<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="debug" default="true"/>
  <arg name="yolo_model" default="yolov11n.pt"/>
  <arg name="input_topic" default="/camera/rgb/image_raw"/>
  <arg name="result_topic" default="/yolo_result"/>
  <arg name="result_image_topic" default="/yolo_image"/>
  <arg name="conf_thres" default="0.50"/>
  <arg name="iou_thres" default="0.45"/>
  <arg name="max_det" default="300"/>
  <arg name="classes" default=""/>
  <arg name="tracker" default="my_track_config.yaml"/>
  <arg name="device" default=""/>
  <arg name="result_conf" default="true"/>
  <arg name="result_line_width" default="1"/>
  <arg name="result_font_size" default="1"/>
  <arg name="result_font" default="Arial.ttf"/>
  <arg name="result_labels" default="true"/>
  <arg name="result_boxes" default="true"/>

  <arg name="camera_info_topic" default="/camera/rgb/camera_info"/>
  <arg name="lidar_topic" default="/converted_pc"/>
  <arg name="yolo_3d_result_topic" default="/yolo_3d_result"/>
  <arg name="cluster_tolerance" default="0.75"/>
  <arg name="voxel_leaf_size" default="0.1"/>
  <arg name="min_cluster_size" default="1"/>
  <arg name="max_cluster_size" default="10000"/>

  <node name="my_tracker_node" pkg="ultralytics_ros" type="my_tracker_node.py" output="screen">
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="input_topic" value="$(arg input_topic)"/>
    <param name="result_topic" value="$(arg result_topic)"/>
    <param name="result_image_topic" value="$(arg result_image_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="result_conf" value="$(arg result_conf)"/>
    <param name="result_line_width" value="$(arg result_line_width)"/>
    <param name="result_font_size" value="$(arg result_font_size)"/>
    <param name="result_font" value="$(arg result_font)"/>
    <param name="result_labels" value="$(arg result_labels)"/>
    <param name="result_boxes" value="$(arg result_boxes)"/>
    <rosparam param="classes" subst_value="true">$(arg classes)</rosparam>
    <rosparam param="device" subst_value="true">$(arg device)</rosparam>
  </node>

  <node name="my_tracker_with_cloud_node" pkg="ultralytics_ros" type="my_tracker_with_cloud_node" output="screen">
    <param name="camera_info_topic" value="$(arg camera_info_topic)"/>
    <param name="lidar_topic" value="$(arg lidar_topic)"/>
    <param name="yolo_result_topic" value="$(arg result_topic)"/>
    <param name="yolo_3d_result_topic" value="$(arg yolo_3d_result_topic)"/>
    <param name="cluster_tolerance" value="$(arg cluster_tolerance)"/>
    <param name="voxel_leaf_size" value="$(arg voxel_leaf_size)"/>
    <param name="min_cluster_size" value="$(arg min_cluster_size)"/>
    <param name="max_cluster_size" value="$(arg max_cluster_size)"/>
  </node>

  <node name="yolo_3d_transform_node" pkg="ultralytics_ros" type="trans_frame_node.py" output="screen"/>
  
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="map_file" default="/home/andre/ros_docker_ws/catkin_ws/map/map.yaml"/>
  <!-- <arg name="map_file" default="/home/andre/ros_docker_ws/catkin_ws/map/test.yaml"/> -->
  <arg name="open_rviz" default="true"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="initial_pose_a" default="0.0"/>
  <arg name="use_teb" default="true"/>
  
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

<!-- For Turtlebot3 Sim -->
  <include file="$(find my_navigation)/launch/amcl.launch">
    <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
  </include>

  <include file="$(find my_navigation)/launch/move_base.launch">
    <arg name="model" value="$(arg model)" />
    <arg name="move_forward_only" value="$(arg move_forward_only)"/>
    <arg name="use_teb" value="$(arg use_teb)"/>
  </include>

<!-- For Tracer Sim -->
  <!-- <include file="$(find my_navigation)/launch/amcl_trace.launch"/>
  <include file="$(find my_navigation)/launch/move_base_trace.launch">
    <arg name="model" value="$(arg model)" />
    <arg name="move_forward_only" value="$(arg move_forward_only)"/>
  </include> -->

  <node if="$(arg debug)" pkg="rviz" type="rviz" name="rviz" args="-d $(find ultralytics_ros)/rviz/my.rviz"/>
</launch>
